File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/valid-sudoku/code/cpp/valid_sudoku_sol1.cpp
class Solution {
public:
    // Three passes: check rows, columns, boxes separately
    bool isValidSudoku(vector<vector<char>>& board) {
        // Check rows
        for (int i = 0; i < 9; i++) {
            unordered_set<char> seen;
            for (int j = 0; j < 9; j++) {
                if (board[i][j] == '.') continue;
                if (seen.count(board[i][j])) return false;
                seen.insert(board[i][j]);
            }
        }
        // Check columns
        for (int j = 0; j < 9; j++) {
            unordered_set<char> seen;
            for (int i = 0; i < 9; i++) {
                if (board[i][j] == '.') continue;
                if (seen.count(board[i][j])) return false;
                seen.insert(board[i][j]);
            }
        }
        // Check 3x3 boxes
        for (int bi = 0; bi < 3; bi++) {
            for (int bj = 0; bj < 3; bj++) {
                unordered_set<char> seen;
                for (int i = bi*3; i < bi*3+3; i++)
                    for (int j = bj*3; j < bj*3+3; j++) {
                        if (board[i][j] == '.') continue;
                        if (seen.count(board[i][j])) return false;
                        seen.insert(board[i][j]);
                    }
            }
        }
        return true;
    }
};

File: src/data/courses/dsa/arrays-and-hashing/frequency-counting/valid-sudoku/code/python/valid_sudoku_sol1.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # Three passes: rows, columns, boxes
        for i in range(9):
            if not self.check([board[i][j] for j in range(9)]): return False
            if not self.check([board[j][i] for j in range(9)]): return False
        for bi in range(3):
            for bj in range(3):
                if not self.check([board[bi*3+i][bj*3+j] for i in range(3) for j in range(3)]): return False
        return True
    def check(self, cells):
        nums = [c for c in cells if c != '.']
        return len(nums) == len(set(nums))